File: scripts/GenomeChronicler_vepTables_fromVEP.pl
#!/usr/bin/env perl

use strict;
use warnings;

use Data::Dumper;

#Sort out input
die "Usage: $0 VEPhtml OutputDir\n" if(@ARGV != 2);
my $filename = $ARGV[0];
my $outDir = $ARGV[1];



die "Major error here (File not found)\n" if(!-e $filename);

open(OUT, ">${outDir}/latest.summary.csv") or die "Could not open output file: $!\n";
print OUT "Feature\tCount\n";

open(IN, "grep \"gen_stats\" $filename |") or die "Could not open input file: $!\n";
my $recording = 0;
while (my $line = <IN>) {
    chomp($line);
    next if($line eq "");
    
    #Note for future generations: This is as silly as Perl gets, but it does the job (for now). I will re-write this with proper parsing at some point, or even better, use an off-the-shelve HTML parsing module.
    
    $line =~ s/<tr>/\n/g;
    $line =~ s/<\/table>/\n<\/table>\n/g;
    $line =~ s/<td>/\t/g;
    $line =~ s/<\/td>|<\/tr>//g;
    
    my @lines = split("\n",$line);
    for my $line (@lines) {
        if($line =~ m/stats_table/) {
            $recording += 1;
            next;
        }
        elsif($line =~ m/\<\/table\>/ and $recording == 2) {
            $recording = 0;
        }
        
        $line =~ s/^\s+//;
        $line =~ s/\s+$//;
        $line =~ s/\s{2,}/\t/g;
        
        next if($line =~ m/Variants processed/);
        next if($line =~ m/Lines of output written/);
        
        print OUT "$line\n" if($recording == 2);
    }
}

close IN;
close OUT;

#system("ln -sf ${filename}.summary.csv latest.summary.csv");


open(IN, "cat $filename | grep drawTable | grep \"\\[\" | ") or die "Could not open input file: $!\n";

while (my $line = <IN>) {
	chomp($line);
	next if($line eq "");

    my ($tabName,$tabString);
    
    if($line =~ m/drawTable\(\'(.+?)\'.+?\[(.+)\]/) {
        $tabName = $1;
        $tabString = $2;
    }
    
    my $counter = 0;
    my $summer = 0;
    my @table;
    while ($tabString =~ m/\[(.+?)\]/g) {
        my $pair = $1;
        $pair =~ s/[']//g;
        $pair =~ s/\_/ /g;
        
        $pair =~ s/\ //g if($counter == 0); #Remove spaces in title only
        my @split = split(",",$pair);
        push(@table,\@split);
        
        $summer += $split[1] if($counter >0);
        $counter++;
    }
    
    open(OUT, ">${outDir}/latest.${tabName}.csv") or die "Could not open output file: $!\n";
    
    my $head = shift(@table);
    $head->[0] = "Label";
    print OUT join(",",@{$head})."\n";
    
    my $other = 0;
    for my $d (@table) {
        my $perc = sprintf("%0.2f",(($d->[1]/$summer)*100));
        if($perc > 1) {
            $d->[0] = ucfirst($d->[0]);
            $d->[1] = $perc;
            print OUT join(",",@{$d})."\n";
        }
        else {
            $other += $d->[1];
        }
    }
    if($other > 0) {
        my $perc = sprintf("%0.2f",(($other/$summer)*100));
        print OUT "Others,$perc\n";
    }
    close OUT;
    
    #system("ln -sf ${filename}.${tabName}.csv latest.${tabName}.csv");
    
    #die "$tabName\n$tabString\n";
}

close IN;

File: scripts/GenomeChronicler_vepTables_fromVEP.py
#!/usr/bin/env python3

import subprocess
import sys
import re
from pathlib import Path

import fire

def get_vepTables_from_VEP(filename, out_dir="."):

    # # Sort out input
    # if len(sys.argv) != 3:
    #     sys.exit("Usage: {} VEPhtml OutputDir".format(sys.argv[0]))

    # filename = sys.argv[1]
    # out_dir = sys.argv[2]

    # if not os.path.exists(filename):
    #     sys.exit("Major error here (File not found)")

    out_file = open("{}/latest.summary.csv".format(out_dir), "w")
    out_file.write("Feature\tCount\n")

    recording = 0
    # with (filename, "r") as in_file:
    with subprocess.Popen(f'cat {filename} | grep "gen_stats"', shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE) as proc:
        for line in proc.stdout:
            line = line.decode()
            # line = line.strip()
            line = line.rstrip("\n")
            if not line:
                continue

            # Note for future generations: This is as silly as Perl gets, but it does the job (for now). I will re-write this with proper parsing at some point, or even better, use an off-the-shelve HTML parsing module.
            line = line.replace("<tr>", "\n")
            line = re.sub(r"</table>", "\n</table>\n", line)
            line = re.sub(r"<td>", "\t", line)
            line = re.sub(r"</td>|</tr>", "", line)

            lines = line.split("\n")
            for l in lines:
                if "stats_table" in l:
                    recording += 1
                    continue
                elif ("</table>" in l.lower()) and (recording == 2):
                    recording = 0

                # l = l.strip()
                l = re.sub(r"^\s+", "", l)
                l = re.sub(r"\s+$", "", l)
                l = re.sub(r"\s{2,}", "\t", l)

                if ("variants processed" in l.lower()):
                    continue
                if ("lines of output written" in l.lower()):
                    continue



                if recording == 2:
                    out_file.write("{}\n".format(l))

    # in_file.close()
    out_file.close()

    with open(filename, "r") as in_file:
        for line in in_file:
            line = line.strip()
            if not line or "drawTable" not in line or "[" not in line:
                continue

            match = re.search(r"drawTable\('(.+?)'.+?\[(.+)\]", line)
            if match:
                tab_name = match.group(1)
                tab_string = match.group(2)

                counter = 0
                summer = 0
                table = []
                for pair in re.findall(r"\[(.+?)\]", tab_string):
                    pair = pair.replace("'", "")
                    pair = pair.replace("_", " ")
                    if counter == 0:
                        pair = pair.replace(" ", "")
                    split = pair.split(",")
                    table.append(split)

                    if counter > 0:
                        summer += int(split[1])

                    counter += 1

                out_file = open("{}/latest.{}.csv".format(out_dir, tab_name), "w")

                head = table.pop(0)
                head[0] = "Label"
                out_file.write("{}\n".format(",".join(head)))

                other = 0
                for d in table:
                    perc = f"{round(float(d[1]) / summer * 100, 2):0.2f}"
                    if float(perc) > 1:
                        d[0] = d[0].capitalize()
                        d[1] = perc
                        out_file.write("{}\n".format(",".join([str(x) for x in d])))
                    else:
                        other += int(d[1])

                if other > 0:
                    perc = f"{round(float(other) / summer * 100, 2):0.2f}"
                    out_file.write(f"Others,{perc}\n")
                    # out_file.write("{}\n".format(",".join([str(x) for x in d])))
                   


if __name__ == '__main__':
    fire.Fire(get_vepTables_from_VEP)
